### PR TITLE
Time improvements

### PR DESCRIPTION
**Pull request by doomlaur**
As promised, this pull request does the following:

1. Moves the wrappers for the C time functions to a common file (they no longer exist twice).
2. Uses `localtime_r` on Linux and other Unix systems that support it to improve thread-safety.
3. Fixes the unit test for weekdays under MSVC.

Note: the improved wrappers for the C time functions are based on the previous ones, but contain the improvements stated above. However, this also means that the new wrappers - just like the old ones - don't have any handling for failed calls to `localtime` or `gmtime`. For example, under MSVC, the [C Standard Library of MSVC has the limitation](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/localtime-localtime32-localtime64?view=msvc-170) for `localtime`, `gmtime` and `mktime` that it only supports dates between January 1st 1970 00:00:00 UTC and December 31st 3000 23:59:59 UTC. However, Excel supports dates beginning with January 1st, 1900, while XLNT under MSVC would not handle it correctly, without any information to the caller of our safe functions that something failed. We could change that by returning an `xlnt::optional` in case we detect that something has failed. Currently, the following will happen:

- for `localtime_s` and `gmtime_s` (under MSVC), as well as `localtime_r` and `gmtime_r` under Unix, a default-initialized `std::tm` would be returned (that is not useful in most cases, of course)
- for `std::localtime` and `std::gmtime` (on other systems than Windows and Unix), ~~we would dereference a `nullptr` and cause a crash~~ *changed to no longer cause a crash and behave like on Windows and Unix - but this still means that the caller has no idea that the call failed*

I definitely think that this is not great and should be improved - but I created this pull request to first ask what you @m7913d think about my proposed solution :slightly_smiling_face:

### CODE REVIEW

File: source/detail/time.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+namespace xlnt {
+namespace detail {
+
+inline std::tm localtime_safe(std::time_t raw_time)
+{
+    std::tm result{};
+
+#ifdef _MSC_VER
+    localtime_s(&result, &raw_time);
+#elif _POSIX_C_SOURCE >= 1 || _XOPEN_SOURCE || _BSD_SOURCE || _SVID_SOURCE || _POSIX_SOURCE
```
Comment (m7913d) [inline source/detail/time.hpp:37]: What is the reason for using those constants? Shouldn't we use the `defined` operator in these checks? In [source/utils/path.cpp](https://github.com/xlnt-community/xlnt/blob/0b772584c11fc342dba35401ac53e85bf41cb654/source/utils/path.cpp#L32) `__linux` and `__APPLE__` are used. Should we use the same here?

File: source/detail/time.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    localtime_r(&raw_time, &result);
+#else
```
Comment (m7913d) [inline source/detail/time.hpp:39]: Maybe we can also check for `__STDC_LIB_EXT1__` and use `localtime_s` in that case: see https://en.cppreference.com/w/c/chrono/localtime. Not sure if we really need this extra case, given most platforms are supported by the two cases above.

File: source/utils/date.cpp [hunk 3]
```diff
@@ -120,7 +105,7 @@ int date::to_number(calendar base_date) const
 
 date date::today()
 {
-    std::tm now = safe_localtime(std::time(nullptr));
+    std::tm now = detail::localtime_safe(std::time(nullptr));
```
Comment (m7913d) [inline source/utils/date.cpp:108]: Some particular reason why you switched `safe_localtime` to `localtime_safe`?

File: source/utils/date.cpp [hunk 4]
```diff
@@ -132,7 +117,7 @@ int date::weekday() const
     tm.tm_year = year - 1900;
     std::time_t time = std::mktime(&tm);
 
-    return safe_localtime(time).tm_wday;
+    return detail::localtime_safe(time).tm_wday;
```
Comment (m7913d) [inline source/utils/date.cpp:120]: Can't we just use `tm.tm_wday`? According to the specs, the `tm` structure will be updated on success: see https://en.cppreference.com/w/cpp/chrono/c/mktime.

File: source/utils/date.cpp [hunk 5]
```diff
@@ -132,7 +117,7 @@ int date::weekday() const
     tm.tm_year = year - 1900;
     std::time_t time = std::mktime(&tm);
```
Comment (m7913d) [inline source/utils/date.cpp:118]: Should we verify if `mktime` succeeded? If not, we could e.g. return -1 too to indicate failure.

**Review by m7913d — COMMENTED** [review]
Concerning error handling: I would be fine with returning a `xlnt::optional` object by your new `*_safe` functions.
However, the only function that reasonably may fail is `date::weekday` (as the others use the current date/time as input), but I think this function doesn't need to use `localtime_safe`.

File: source/detail/time.hpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2024 xlnt-community
```
[the remark this answers is not in the document]
Comment (doomlaur) [reply source/detail/time.hpp:1]: I was honestly a bit unsure how to call it - whether it should be a general file for OS-specific wrappers, or something time-specific. However, to keep the number of includes to a minimum and to group the functions in a better way, I decided to create a file for time-specific functions (date, time, date/time) and simply called it `time.hpp`. But I'm open for suggestions :wink: We could also call it something like `time_wrappers.hpp` - on the other hand, all the other filenames in the same directory are kept rather short.

[the remark this answers is not in the document]
Comment (m7913d) [reply source/detail/time.hpp:1]: I'm just wondering if the same filename could give some trouble on certain platforms. If not, we can keep time.hpp, otherwise time_wrappers, or more general time_helpers or time_conversion may be better.

File: source/detail/time.hpp (hunk 1 above)
Comment (doomlaur) [reply source/detail/time.hpp:37]: I took the feature test macros from here: https://linux.die.net/man/3/localtime_r

The thing is, in my experience, the more you go into platform-specific functions, the harder it gets - and, unfortunately, this often goes beyond differentiating between Windows, Linux and macOS. Because I just looked over miniz yesterday - look what they are doing in order to support all the different platforms: https://github.com/richgel999/miniz/blob/master/miniz_zip.c#L104

My point is: I want to use the new code to improve code quality and safety - however, due to the code being less portable, we need to be much more careful to still compile on different platforms. Since I did not find any better documentation on the feature test macros, I think the current solution is the best way to go to in order to ensure maximum portability :slightly_smiling_face: 

Some examples I heard of but did not experience them myself: Android is Linux, but Android still has its own SDK which might be different in some ways. Also, there are some differences between Clang and Apple Clang, so again, just testing for one might lead to wrong results. In the end, the only real way to really make sure it works on as many platforms as possible is to let the community test it, but at least we can try our best to write portable code :slightly_smiling_face:

File: source/utils/date.cpp (hunk 3 above)
Comment (doomlaur) [reply source/utils/date.cpp:108]: Personally, I prefer seeing as soon as possible which function I'm actually calling. In my own projects, I prefer using a suffix like `_safe` instead of a prefix, because this way I can see more easily the difference between e.g. `localtime_safe` and `gmtime_safe`. But this is just my preference :smile: We can do it differently - I don't mind, as long as it's easily understandable by everyone and makes sense!

File: source/utils/date.cpp (hunk 4 above)
Comment (doomlaur) [reply source/utils/date.cpp:120]: You are right, this should work and would avoid an otherwise unnecessary call to `localtime_safe`. Good point :+1:

File: source/utils/date.cpp (hunk 5 above)
Comment (doomlaur) [reply source/utils/date.cpp:118]: I'm not so sure this is a good idea, and I'm honestly surprised that [the documentation on cppreference](https://en.cppreference.com/w/cpp/chrono/c/mktime) states that -1 is a sign of failure. Under MSVC, this is indeed the case, because MSVC cannot use dates before 1970-1-1, and -1 would be 1969-12-31 23:59:59. However, for other platforms, it is often the case that dates at least back to 1900-1-1 are supported, and -1 is simply one second before 1970-1-1.

In other words, I would only trust -1 to be a sign of failure under MSVC. However, I would be interested if the C specification gives any more details on this.

File: source/detail/time.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (doomlaur) [reply source/detail/time.hpp:1]: I really do hope that no compiler and/or build system would work so badly that different paths containing the same filename would cause issues :sweat_smile: What I would rather think about it is that it would be easier for other contributors to XLNT to open the wrong file, causing confusions. This is however already an issue with many other libraries like the C++ STL and Boost (they often have similar filenames), so we are not the first ones having this issue. This is more of a theoretical nature, as I'm not aware of people actually complaining about this. But if you prefer me to change the name, I'll do it, of course :wink:

File: source/detail/time.hpp (hunk 2 above)
Comment (m7913d) [reply source/detail/time.hpp:39]: Thanks for the clarification. If nobody implemented the standardized `localtime_s` version, it's indeed useless to add it as a fallback.

File: source/utils/date.cpp (hunk 3 above)
Comment (m7913d) [reply source/utils/date.cpp:108]: Ok, fine for me.

File: source/utils/date.cpp (hunk 5 above)
Comment (doomlaur) [reply source/utils/date.cpp:118]: Apparently, [the C23 specification](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3088.pdf) still states that `If the calendar time cannot be represented, the function returns the value (time_t) (-1)`. I'm seriously confused by this and seriously doubt that this is actually the case on any other compilers than MSVC. Unless, of course, this still happens although -1 could be a valid value, in which case it's not possible on such compilers to differentiate between the actual second -1 being 1969-12-31 23:59:59 and an actual error case.

File: source/utils/date.cpp (hunk 4 above)
Comment (doomlaur) [reply source/utils/date.cpp]: I added a commit that contains your proposed improvement - thanks! :+1:

File: source/detail/time.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (m7913d) [reply source/detail/time.hpp:1]: Batch/inference rules in NMake may be an example (at least if generated by qmake for source files): see https://bugreports.qt.io/browse/QTVSADDINBUG-424, but that's maybe irrelevant for our use case. So, leave it as you like.

File: source/utils/date.cpp [hunk 7]
```diff
@@ -105,8 +105,16 @@ int date::to_number(calendar base_date) const
...
-    return date(1900 + now.tm_year, now.tm_mon + 1, now.tm_mday);
+    optional<std::tm> now = detail::localtime_safe(std::time(nullptr));
+
+    if (now.is_set())
+    {
+        return date(1900 + now.get().tm_year, now.get().tm_mon + 1, now.get().tm_mday);
+    }
+    else
+    {
+        return date(1970, 1, 1);
```
Comment (m7913d) [inline source/utils/date.cpp:116]: Is this the best "invalid" date? Maybe we should use a real invalid date like `date(0,0,0)`? We can even add a `is_valid` or `is_null` member to `xlnt::date` to make it clear.

File: source/detail/time.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (doomlaur) [reply source/detail/time.hpp]: Alright, I renamed it to `time_helpers.hpp` :+1:

File: source/utils/date.cpp (hunk 5 above)
Comment (m7913d) [reply source/utils/date.cpp]: Maybe the solution provided in [this doc](https://pubs.opengroup.org/onlinepubs/9799919799/functions/mktime.html) is a good solution:

> Since (time_t)-1 is a valid return value for a successful call to mktime(), an application wishing to check for error situations should set tm_wday to a value less than 0 or greater than 6 before calling mktime(). On return, if tm_wday has not changed an error has occurred.

File: source/utils/date.cpp (hunk 7 above)
Comment (doomlaur) [reply source/utils/date.cpp:116]: I didn't like this solution either, and I think I should really change it right away. The only reason why I hesitated is because setting proper time limits is, unfortunately, a mess - thus, setting a default date is also a mess. In my experience:
1. Using anything outside of the years [0; 9999] causes crashes under MSVC depending on what you're doing (e.g. formatting dates using `std::format`). 
2. Using anything outside of the years [1970; 3000] (or [1970; 2038] under 32-bit) causes issues under MSVC.
3. Other platforms have other limits currently unknown to me (I could try it in the Compiler Explorer / Godbolt, though).

Any such solution would lead to some magic number (or magic date, in this case), which is definitely not nice either.

I think the only proper way to do it would be to have a function like `is_valid`. For example, we could allow default-constructing an `xlnt::date` object that would be invalid by default.

Comment (m7913d) [reply source/utils/date.cpp:116]: Ok, for adding default constructor for invalid date. The only reason why I would prefer `is_null` instead of `is_valid` is that date(2000, 13, 35) is invalid too, but a simple `is_valid` check (e.g. year == date == month == 0) wouldn't say so.

File: source/utils/date.cpp (hunk 5 above)
Comment (doomlaur) [reply source/utils/date.cpp]: That solution is a bit cursed, but I like it better than no error handling at all. Thanks! :smile:

File: source/utils/date.cpp (hunk 7 above)
Comment (doomlaur) [reply source/utils/date.cpp]: My newest commit contains your proposed changes. Let me know if you have any other suggestions!

File: include/xlnt/utils/date.hpp [hunk 8]
```diff
@@ -61,8 +68,9 @@ struct XLNT_API date
     /// <summary>
     /// Calculates and returns the day of the week that this date represents in the range
     /// 0 to 6 where 0 represents Sunday.
+    /// Returns an empty optional if the weekday could not be determined.
     /// </summary>
-    int weekday() const;
+    optional<int> weekday() const;
```
Comment (m7913d) [inline include/xlnt/utils/date.hpp:73]: This changes a public interface. I don't think we should to this with a minor release. I would just return -1 in case of a failure, at least for now.

File: include/xlnt/utils/date.hpp [hunk 9]
```diff
@@ -49,10 +51,15 @@ struct XLNT_API date
...
     /// <summary>
-    /// Constructs a data from a given year, month, and day.
+    /// Constructs a date from a given year, month, and day.
     /// </summary>
     date(int year_, int month_, int day_);
 
+    /// <summary>
+    /// Constructs an empty date (a call to is_null will return true).
+    /// </summary>
+    date() = default;
```
Comment (m7913d) [inline include/xlnt/utils/date.hpp:61]: I'm not sure if I like the default constructor here as it will result in a random integer as year/month/day.

File: include/xlnt/utils/datetime.hpp [hunk 10]
```diff
@@ -131,6 +135,11 @@ struct XLNT_API datetime
     /// The microsecond
     /// </summary>
     int microsecond;
+
+    /// <summary>
+    /// Whether the date is in an empty state.
+    /// </summary>
+    bool is_null = true;
```
Comment (m7913d) [inline include/xlnt/utils/datetime.hpp:142]: Using a boolean for null state just increases class size unnecessarily. I think using a magic 0/0/0 as day/month/year is fine, as long as this magic number is internal to the class.

File: source/utils/date.cpp [hunk 11]
```diff
@@ -26,10 +26,14 @@
 #include <xlnt/utils/date.hpp>
 #include <detail/time_helpers.hpp>
 
+/// Invalid weekday for checking whether std::mktime was successful - see below.
+/// Must be outside of the range [0, 6].
+constexpr int INVALID_WDAY = -1;
```
Comment (m7913d) [inline source/utils/date.cpp:31]: Should we add this to an unnamed namespace, to ensure we never violate the one definition rule?

File: tests/utils/datetime_test_suite.cpp [hunk 12]
```diff
@@ -117,12 +117,12 @@ class datetime_test_suite : public test_suite
 
     void test_weekday()
     {
-        xlnt_assert_equals(xlnt::date(2000, 1, 1).weekday(), 6); // January 1st 2000 was a Saturday
-        xlnt_assert_equals(xlnt::date(2016, 7, 15).weekday(), 5); // July 15th 2016 was a Friday
-        xlnt_assert_equals(xlnt::date(2018, 10, 29).weekday(), 1); // October 29th 2018 was Monday
+        xlnt_assert_equals(xlnt::date(2000, 1, 1).weekday().get(), 6); // January 1st 2000 was a Saturday
```
Comment (m7913d) [inline tests/utils/datetime_test_suite.cpp:120]: I don't like changes to unit tests, as this also means necessary refactoring in production code.

**Review by m7913d — CHANGES_REQUESTED** [review]

File: include/xlnt/utils/date.hpp (hunk 8 above)
Comment (doomlaur) [reply include/xlnt/utils/date.hpp]: While I really dislike magic numbers like -1 (often used in old C/C++ functions) as they are often forgotten to be checked, and while I also don't think that `xlnt::date::weekday` and `xlnt::datetime::weekday` are used that widely (or at least so much that a change would be a lot of work), I do agree that releasing a new version of XLNT that only contains bug fixes is more important right now. Thus, I reverted my change and documented that -1 will be returned.

File: include/xlnt/utils/date.hpp (hunk 9 above)
Comment (doomlaur) [reply include/xlnt/utils/date.hpp:61]: While other structs like [`std::chrono::year_month_day`](https://en.cppreference.com/w/cpp/chrono/year_month_day/year_month_day) do the same (unfortunately!), I agree that this is not nice, and it was not my intention. The year/month/day fields are now initialized to 0 by default, after my latest commit.

File: tests/utils/datetime_test_suite.cpp (hunk 12 above)
Comment (doomlaur) [reply tests/utils/datetime_test_suite.cpp]: Due to the reverted changes (see my other comment), the unit tests are no longer changed. However, I would highly prefer to change it to an `xlnt::optional` instead of using magic numbers like -1 for the next major release.

File: include/xlnt/utils/date.hpp (hunk 8 above)
Comment (doomlaur) [reply include/xlnt/utils/date.hpp]: However, I would highly prefer to change it to an `xlnt::optional` instead of using magic numbers like -1 for the next major release.

File: source/utils/date.cpp (hunk 11 above)
Comment (doomlaur) [reply source/utils/date.cpp:31]: Could this really ever be an ODR violation? To my knowledge, this variable should only have internal linkage, thus it should only be available in the translation unit of `date.cpp` (note: not a header file!). Personally, I never had issues with such variables in my project, but feel free to prove me wrong :sweat_smile:

File: include/xlnt/utils/datetime.hpp (hunk 10 above)
Comment (doomlaur) [reply include/xlnt/utils/datetime.hpp:142]: To be precise, it increases `xlnt::date` by 4 bytes (1 byte for the boolean and 3 bytes for padding) to a total of 16 bytes. Also, it increases `xlnt::datetime` by 4 bytes (1 boolean, 3 padding) to a total of 32 bytes.

However, I personally disagree that it is not necessary. Having a magic number for that purpose - even if it is internal - means that a user of `xlnt::date` or `xlnt::datetime` would never be allowed to use the magic date, as it would mean that the value is invalid. Even worse, since this is internal, the user would never have the possibility to check it either, but would be required to read the internals of our classes.

While writing correct code would be possible without an additional variable, I think our `xlnt::date` and `xlnt::datetime` have not really been made for that (not yet, at least). Examples from C++20:

- [`std::chrono::day` has an `ok()`](https://en.cppreference.com/w/cpp/chrono/day/ok) function that checks whether a day is in the allowed range
- [`std::chrono::month` has an `ok()`](https://en.cppreference.com/w/cpp/chrono/month/ok) function that checks whether a month is in the allowed range
- [`std::chrono::year` has an `ok()`](https://en.cppreference.com/w/cpp/chrono/year/ok) function that checks whether a year is in the allowed range
- [`std::chrono::year_month_day` has an `ok()`](https://en.cppreference.com/w/cpp/chrono/year_month_day/ok) that checks whether the year, month and day are in range (like above), and further checks things like leap years (whether February 29 is valid in the specified year) and whether 31st of a month is allowed in the specified month.

The new C++20 structures allow this **_without an additional variable_** by specifying very precise rules on how to handle out of range values. However, we're only using `int`s for all 3 fields, and we're not enforcing any ranges or checking them at all. In other words, we currently have no validation, so we also cannot do any sanity checks. A magic number would make the situation even worse.

Besides, if we really wanted to decrease the class size we could use `unsigned char` / `uint8_t` for the month, day, hour, minute, second. With an adapted order of the fields (higher-sized types at the top - that is, year and microsecond) that would result in a class size of 16 bytes for `xlnt::datetime` and a size of 8 for `xlnt::date`.

Comment (m7913d) [reply include/xlnt/utils/datetime.hpp:141]: Ok, let's keep it like this for now.

File: source/utils/date.cpp (hunk 11 above)
Comment (m7913d) [reply source/utils/date.cpp:31]: I often use unnamed namespaces anyway, so I don't need to keep in mind all the default linkage rules. But indeed, not needed for a const expr.

File: include/xlnt/utils/datetime.hpp (hunk 10 above)
Comment (m7913d) [reply include/xlnt/utils/datetime.hpp:141]: Another thing that I do like even less is that we expose this `is_null` member publicly. (sorry that I didn't catch it on first review). I know the other members are also public (and I think we should change this in the next major release), but storing is_null as a boolean really seems to me as an implementation detail.
What do you think about changing this to a private member and provide getter/setter functions?
Another inconvenience is that after setting year/month/day the user has also to explicitly set the `is_null` to false. Maybe we should already add setter/getter functions for year/month/day and automatically set is_null to false (and already deprecate access to the public member variables).

File: source/utils/datetime.cpp
Comment (m7913d) [inline source/utils/datetime.cpp:1]: Not all functions take the is_null state into account. Even the constructor `datetime(int year_, int month_, int day_, int hour_, int minute_, int second_, int microsecond_)` does not correctly set the is_null state.

Comment (doomlaur) [reply source/utils/datetime.cpp:1]: You are right - I forgot that constructor, but my newest commit fixed that issue. However, I would like to discuss a bit about your statement `Not all functions take the is_null state into account.`

You are of course right - however, the question that we should ask ourselves is what should happen when having an invalid state (that is, `is_null()` is `true`). Unless we would use a more complicated solution (e.g. having a `create()` function that returns a valid `unique_ptr` on success or `nullptr` on failure), we'll have to deal with an invalid state (and thus checking `is_null()`) in some cases. I tried to avoid using an empty state where it could be easily avoided, however, doing the same for the other functions is much harder. Examples:

- What should happen when calling e.g. `get_year()` on an invalid `date`? Should an exception be thrown? Probably not.
- What should happen when calling `to_number()` on an invalid `datetime`? Should `NaN` be returned?
- What should happen when calling `to_string()` on an invalid `datetime`? Should an empty string be returned? Should the string "invalid" be returned?

Without changing everything to return `optional`, these questions are rather hard to answer. Either way, the way we handle it at the moment is unfortunately pretty error-prone (since the programmer MUST NOT forget to call `is_null()` in case the state could be invalid), but changing that without modifying the public interface is hard. Changing the public interface (which is something we could think about for a future release) would make the handling much more correct (e.g. not handling an `optional` correctly is programmer's own fault), however, this would also make it harder to use the structs as well, since further checks like `is_set()` would be required everywhere.

These are unfortunately a lot of questions that are hard to answer correctly - however, I'm open for dicsussions on what we can do to improve our interface as it is now. Of course, feel free to also express ideas about future developments.

File: include/xlnt/utils/datetime.hpp (hunk 10 above)
Comment (doomlaur) [reply include/xlnt/utils/datetime.hpp]: I agree, and I dislike it as well. My newest commit has changed that and has also deprecated directly accessing the member variables (for now only as a code comment).

I agree to your other remarks. Since I'm a big fan of simple solutions whenever they are possible (as long as they are correct, obviously), I would propose the following solution that I implemented as part of my newest commit:

1. `xlnt::date` and `xlnt::datetime` are no longer meant for direct access to their member variables (which is still allowed for now, but using the getters is encouraged). From now on, they are no longer meant to be modified on the fly: changing an `xlnt::date` or `xlnt::datetime` should only be done by constructing a new one, which is very cheap performance-wise. This whole idea makes a lot of things much easier, as we only have to handle the initial state correctly, but state can no longer be changed on the fly. This also avoids us having to provide setters. It is similar to [`std::chrono::year_month_day`](https://en.cppreference.com/w/cpp/chrono/year_month_day), since it also only has getters but no setters.
2. Both `xlnt::date` and `xlnt::datetime` can now be default-constructed, in which case they are initialized with an empty/invalid state.
3. Both `xlnt::date` and `xlnt::datetime` can be initialized using integers for the year/month/day etc. - however, like before, we still perform no checks on the members, but assume that users know what they are doing and simply assume that the state is NOT empty/invalid.

So, in short, we can either have default-constructed instances having empty/invalid states, or programmer-constructed instances always having valid states. Also, the whole state can no longer change after construction (or should not beginning with the next major release). This makes the code much simpler.

Comment (m7913d) [reply include/xlnt/utils/datetime.hpp]: I'm fine with a const date(time) object.

File: source/utils/datetime.cpp
Comment (m7913d) [reply source/utils/datetime.cpp:1]: Personally, I think throwing an exception may be the right thing to do, as those functions shouldn't be called on an invalid date(time). This is similar to `xlnt::optional::get()`.
As an exception, `to_string` may return an empty string.

For future development, the most important thing is to get rid of the public member variables. Full date manipulation features would be nice, but not my may concern for this library.

Comment (doomlaur) [reply source/utils/datetime.cpp:1]: Agreed - I implemented the requested changes in my latest commit. Furthermore, I tried to document the functions a little better by recommending the user to check `is_null()`, and specified when exceptions are thrown.

Note: `xlnt::date::to_number()` returns an `int`, and since it is not possible to signal an invalid state to the caller (except by using an `optional<int>`, which we do not want to change until the next major version), I decided to throw an exception in that case. In case of `xlnt::datetime::to_number()`, since we're returning a `double`, I could have simply returned `std::numeric_limits<double>::quiet_NaN()` - however, I preferred to keep consistency with `xlnt::date::to_number()`, so I'm currently throwing an exception in that case as well.

File: include/xlnt/utils/date.hpp [hunk 13]
```diff
@@ -49,21 +50,54 @@ struct XLNT_API date
...
+    /// <summary>
+    /// Return the number of days between this date and base_date. The date could be in an empty/invalid state, so you may want to call is_null() first!
+    /// If this function is called when having an empty/invalid state, an exception will be thrown.
     /// </summary>
     int to_number(calendar base_date) const;
 
     /// <summary>
     /// Calculates and returns the day of the week that this date represents in the range
     /// 0 to 6 where 0 represents Sunday.
+    /// Returns -1 if the weekday could not be determined.
```
Comment (m7913d) [inline include/xlnt/utils/date.hpp:71]: If you would like to avoid the magic number, we could also throw an exception instead, e.g. `xlnt::unsupported` . However, I'm not sure which option I prefer, so you can choose freely.

File: source/utils/date.cpp [hunk 14]
```diff
@@ -96,6 +85,11 @@ bool date::operator!=(const date &comparand) const
 
 int date::to_number(calendar base_date) const
 {
+    if (_is_null)
+    {
+        throw xlnt::exception("cannot convert invalid/empty date to a number");
```
Comment (m7913d) [inline source/utils/date.cpp:90]: Most of the time a derived class is thrown. `xlnt::invalid_attribute` may be appropriate.

File: include/xlnt/utils/date.hpp (hunk 13 above)
Comment (doomlaur) [reply include/xlnt/utils/date.hpp:71]: Personally, for my own C++ projects, I prefer to avoid exceptions whenever possible, because there's almost always a better solution (except if you work with some library that requires exception handling). Also, in performance-critical code, catching exceptions frequently reduces performance considerably (probably not important here, but it's worth to mention). An alternative to exception handling could either be a `bool` return value specifying whether the value is valid, or even better, `std::optional` if C++17 is available (or, in our case, `xlnt::optional`). I prefer instead to do the following:

- Use exceptions, if possible, only when having absolutely unrecoverable errors (together with an unhandled exception handler).
- Otherwise, use exceptions only if required by some library and catch the exception as soon as possible, WITHOUT propagating it.
- Otherwise use `bool`s or `optional`s for signaling error cases. For more complex errors, I prefer `std::variant` (to have either a result or an error) if C++17 support is available, or even better, `std::expected` if C++23 support is available.

Also, if stated very clearly, I also have nothing against using exceptions if the programmer messed up badly, e.g. accessed the value of an `std::optional` or `xlnt::optional`, or accessed invalid indices when using `std::vector::at()`. This effectively means that it is NOT necessary to write `try` / `catch` all over the code in order to write correct code. I think this is important because C++ does not enforce handling all exceptions when calling a function that could throw, so it requires everyone to read the documentation (if it exists at all; and even then, many people unfortunately don't read the documentation).

However, I do agree that in this specific case, having a behavior similar to `std::optional` or `xlnt::optional` in the sense that access to invalid values is prevented using exceptions is definitely much better than allowing unrestricted access to invalid values. However, what I personally want to avoid is having to write `try` / `catch` all over the code due to things the programmer cannot control, e.g. a failed internal call to `detail::localtime_safe` or `std::mktime`.

In conclusion, while I really dislike using a magic number, using an exception in this case provides no advantages but a huge disadvantage, since it requires exception handling everywhere, making the situation even worse.

The only proper solution, in my opinion, is to return an `xlnt::optional` - this way, the programmer can see right away that the call can fail, even without reading the documentation. This also applies to `date::today()`, `datetime::today()` and `datetime::now()` (although, one could argue that it's unlikely to not be able to get today's date and/or time, so maybe using an `optional` here would be exaggerated). And now I even got to a point where I can argue that, when returning an `optional` in this case(s), our internal variable `_is_null` is not even necessary anymore, as the structs will always only contain a valid state (at least in the sense that we're not checking any ranges for the year, month, day etc.).

Of course, all of this is not possible without changing a public interface, so not doable if we want to release a minor version.

Either way, I still think that getting rid of the public members in favor of the getters is a very good idea (e.g. so that the struct can only be changed during creation). I only explained the whole thing above to prove why I think that a slightly different solution would be better, but our current solution is still good enough for now, I think.

File: source/utils/date.cpp (hunk 14 above)
Comment (doomlaur) [reply source/utils/date.cpp:90]: I agree. For better exception handling using `try` / `catch`, using a derived exception class would be better. However, at the moment, using `xlnt::exception` allows me to set an exception message, while using `xlnt::invalid_attribute` does not. Since I don't like using `try` / `catch` all over my code (see my long explanation above), I would personally prefer having a proper error message instead of a derived exception class - but I know that other programmers might disagree with me.

We would have the following options:

1. Adapt `xlnt::invalid_attribute` to accept a custom error message, meaning that both advantages could be achieved at once.
2. Switch to `xlnt::invalid_attribute` for a better `try` / `catch` experience, at the cost of a worse error message.
3. Leave the exceptions how they are at the moment, with good error messages, but not so great `try` / `catch` experience.

Which one of the options should we choose? :slightly_smiling_face: I would personally prefer option 1 (while keeping the current default message, if no custom one is specified), but I'm open for discussion :wink: 

Note: I just pushed a commit with even better error messages when throwing exceptions - however, I can gladly switch to any other option.

File: include/xlnt/utils/date.hpp (hunk 13 above)
Comment (m7913d) [reply include/xlnt/utils/date.hpp:71]: Ok. This rule of thumb is fine for me.

Removing `_is_null` in favor of return an `xlnt::optional` may be indeed an option for a major upgrade. In that case, it may be useful to make the constructor private too and use a factory method so we can signal a failed construction too (if desired).

Comment (m7913d) [reply include/xlnt/utils/date.hpp:71]: Maybe we should already create a v2.0.0 milestone so we can keep track of the changes we want to implement once we release a new major version.

File: source/utils/date.cpp (hunk 14 above)
Comment (m7913d) [reply source/utils/date.cpp:90]: Option 1 is fine for me.

**Comment by m7913d** [comment]
As a minor remark, coverage is decreased due to the `is_null` conditions not being unit-tested (ensure exception is thrown, resulting string is empty).

File: include/xlnt/utils/date.hpp (hunk 13 above)
Comment (doomlaur) [reply include/xlnt/utils/date.hpp:71]: Agreed, I think this would be a great idea :+1:

File: source/utils/date.cpp (hunk 14 above)
Comment (doomlaur) [reply source/utils/date.cpp]: My last commit contains your suggested improvements :wink:

**Comment by doomlaur** [comment]
Good point, thank you! I added unit tests for covering all cases that might fail due to an invalid `xlnt::date` or `xlnt::datetime`.

**Comment by doomlaur** [comment]
And, once again, CI has caught a case where all unit tests passed on MSVC, but failed on Linux. Nice! :smile:

File: include/xlnt/utils/date.hpp [hunk 15]
```diff
@@ -49,21 +50,54 @@ struct XLNT_API date
...
     date(int year_, int month_, int day_);
 
     /// <summary>
-    /// Return the number of days between this date and base_date.
+    /// Constructs an empty date (a call to is_null() will return true).
+    /// </summary>
+    date() = default;
+
+    /// <summary>
+    /// Return the number of days between this date and base_date. The date could be in an empty/invalid state, so you may want to call is_null() first!
```
Comment (m7913d) [inline include/xlnt/utils/date.hpp:63]: As we are changing this line, it may be a good time to fix the grammar => return**s**?

File: include/xlnt/utils/date.hpp [hunk 16]
```diff
@@ -49,21 +50,54 @@ struct XLNT_API date
...
     /// <summary>
     /// Calculates and returns the day of the week that this date represents in the range
     /// 0 to 6 where 0 represents Sunday.
+    /// Returns -1 if the weekday could not be determined.
     /// </summary>
     int weekday() const;
 
+    /// <summary>
+    /// Returns the year of the date. The date could be in an empty/invalid state, so you may want to call is_null() first!
+    /// If this function is called when having an empty/invalid state, an exception will be thrown.
```
Comment (m7913d) [inline include/xlnt/utils/date.hpp:77]: Should we specify which exception will be thrown or is that an implementation detail?

File: include/xlnt/utils/exceptions.hpp [hunk 17]
```diff
@@ -240,9 +240,9 @@ class XLNT_API invalid_attribute : public exception
 {
 public:
     /// <summary>
-    /// Default constructor.
+    /// Constructor.
     /// </summary>
-    invalid_attribute();
+    explicit invalid_attribute(const char *optional_message = nullptr);
```
Comment (m7913d) [inline include/xlnt/utils/exceptions.hpp:245]: just wondering, is there any reason why you switched to an explicit constructor?

**Review by m7913d — COMMENTED** [review]
Just some minor remarks. We are almost there.

File: include/xlnt/utils/date.hpp (hunk 15 above)
Comment (doomlaur) [reply include/xlnt/utils/date.hpp]: I fixed the grammar there and in a few other places related to date/time functions :+1:

File: include/xlnt/utils/date.hpp (hunk 16 above)
Comment (doomlaur) [reply include/xlnt/utils/date.hpp]: Usually I would say no, because the programmer should call `is_null()` first, instead of relying on exception handling. However, this is just how I personally see it, and other people might have other preferences. Due to that, I changed the documentation to specify the exact exception being thrown.

**Review by m7913d — APPROVED** [review]
Thanks for all the changes!

File: include/xlnt/utils/exceptions.hpp (hunk 17 above)
Comment (doomlaur) [reply include/xlnt/utils/exceptions.hpp:245]: Since this is now a constructor with one parameter, I would like to avoid the compiler from allowing implicit conversions. To be precise, I would like to avoid [such issues](https://stackoverflow.com/a/121216).
